wavebox: 10.124.17-2 -> 10.127.10-2

flexiondotorg opened the pull request:
## Description of changes

Update Wavebox to 10.127.10-2 and align packaging with Google Chrome.

- Since Wavebox 10.x, Wavebox is built around Chromium, not Electron.
  - This patch updates the Wavebox packaging and is largely based on Google Chrome in Nixpkgs.
- Fixes GL acceleration.
- Improves Wayland support.
- Updates version checker.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>wavebox</li>
  </ul>
</details>

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

flexiondotorg committed:
wavebox: 10.124.17-2 -> 10.126.14-2

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/wavebox/default.nix:
```diff
@@ -1,94 +1,153 @@
-    ln -s $out/opt/wavebox/product_logo_128.png $out/share/icons/hicolor/128x128/apps/wavebox.png
+
+    substituteInPlace $out/share/wavebox.io/$appname/$appname-launcher \
+      --replace 'CHROME_WRAPPER' 'WRAPPER'
+    substituteInPlace $out/share/applications/$appname.desktop \
+      --replace /opt/wavebox.io/wavebox/wavebox-launcher $exe
+    substituteInPlace $out/share/menu/$appname.menu \
+      --replace /opt $out/share \
+      --replace $out/share/wavebox.io/$appname/$appname $exe
```
```suggestion
      --replace-fail 'CHROME_WRAPPER' 'WRAPPER'
    substituteInPlace $out/share/applications/$appname.desktop \
      --replace-fail /opt/wavebox.io/wavebox/wavebox-launcher $exe
    substituteInPlace $out/share/menu/$appname.menu \
      --replace-fail /opt $out/share \
      --replace-fail $out/share/wavebox.io/$appname/$appname $exe
```

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/wavebox/default.nix:
```diff
@@ -1,94 +1,153 @@
+    done
+
+    makeWrapper "$out/share/wavebox.io/$appname/$appname" "$exe" \
+      --prefix LD_LIBRARY_PATH : "$rpath" \
```
Setting LD_LIBRARY_PATH is highly discouraged especially with many libraries. Can this be reduced to the bare minimum instead of using a shotcun approach?

flexiondotorg replied:
Thanks for the detailed review and suggestions :bow: 

From v10.x onwards, Wavebox switched from an Electron app to a Chromium-derived browser. There were a number of issues regarding GPU acceleration in the packaging as a result of this change.

I reworked the Wavebox package based on a pretty much carbon copy of the Google Chrome package in Nixpkgs. Doing so resolved all the issues related to Wavebox finding GL-related libraries.

The use of `LD_LIBRARY_PATH` comes from the Google Chrome package. I have not investigated if it is required or if an alternative approach is suitable.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/wavebox/default.nix:
```diff
@@ -1,94 +1,153 @@
+  '';
+
+  rpath = lib.makeLibraryPath deps + ":" + lib.makeSearchPathOutput "lib" "lib64" deps;
+  binpath = lib.makeBinPath deps;
```
```suggestion
```

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/wavebox/default.nix:
```diff
@@ -1,94 +1,153 @@
+
+    makeWrapper "$out/share/wavebox.io/$appname/$appname" "$exe" \
+      --prefix LD_LIBRARY_PATH : "$rpath" \
+      --prefix PATH            : "$binpath" \
```
```suggestion
      --prefix PATH            : "${lib.makeBinPath deps}" \
```

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/wavebox/default.nix:
```diff
@@ -1,94 +1,153 @@
 
-    mkdir -p $out/bin $out/opt/wavebox
-    cp -r * $out/opt/wavebox
+    appname=wavebox
+
+    exe=$out/bin/$appname
```
```suggestion
    exe=$out/bin/wavebox
```

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/wavebox/default.nix:
```diff
@@ -1,94 +1,153 @@
+      logo_output_prefix="$out/share/icons/hicolor"
+      logo_output_path="$logo_output_prefix/''${icon_size}x''${icon_size}/apps"
+      mkdir -p "$logo_output_path"
+      mv "$icon_file" "$logo_output_path/$appname.png"
```
```suggestion
      mv "$icon_file" "$logo_output_path/wavebox.png"
```

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/wavebox/default.nix:
```diff
@@ -1,94 +1,153 @@
+      mv "$icon_file" "$logo_output_path/$appname.png"
+    done
+
+    makeWrapper "$out/share/wavebox.io/$appname/$appname" "$exe" \
```
```suggestion
    makeWrapper "$out/share/wavebox.io/wavebox/wavebox" "$exe" \
```

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/wavebox/default.nix:
```diff
@@ -1,94 +1,153 @@
+      --prefix PATH            : "$binpath" \
+      --suffix PATH            : "${lib.makeBinPath [ xdg-utils ]}" \
+      --prefix XDG_DATA_DIRS   : "$XDG_ICON_DIRS:$GSETTINGS_SCHEMAS_PATH:${addOpenGLRunpath.driverLink}/share" \
+      --set CHROME_WRAPPER  "$appname" \
+      --add-flags "\''${NIXOS_OZONE_WL:+\''${WAYLAND_DISPLAY:+--ozone-platform-hint=auto --enable-features=WaylandWindowDecorations}}" \
+      --add-flags ${lib.escapeShellArg commandLineArgs}
+
+    for elf in $out/share/wavebox.io/$appname/{wavebox,chrome-sandbox,chrome_crashpad_handler}; do
```
```suggestion
      --set CHROME_WRAPPER "wavebox" \
      --add-flags "\''${NIXOS_OZONE_WL:+\''${WAYLAND_DISPLAY:+--ozone-platform-hint=auto --enable-features=WaylandWindowDecorations}}" \
      --add-flags ${lib.escapeShellArg commandLineArgs}

    for elf in $out/share/wavebox.io/wavebox/{wavebox,chrome-sandbox,chrome_crashpad_handler}; do
```

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/wavebox/default.nix:
```diff
@@ -1,94 +1,153 @@
-    qt5.qtbase
+  nativeBuildInputs = [ patchelf makeWrapper ];
+  buildInputs = [
+    # needed for GSETTINGS_SCHEMAS_PATH
+    gsettings-desktop-schemas glib gtk3
```
```suggestion
    gsettings-desktop-schemas # needed for GSETTINGS_SCHEMAS_PATH
    glib
    gtk3
```

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/wavebox/default.nix:
```diff
@@ -1,94 +1,153 @@
+# command line arguments which are always set e.g "--disable-gpu"
+, commandLineArgs ? ""
+
+# Will crash without.
```
```suggestion
```
so it is just another dependency like above

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/wavebox/default.nix:
```diff
@@ -1,94 +1,153 @@
+# Loaded at runtime.
+, libexif, pciutils
+
+# Additional dependencies according to other distros.
+## Ubuntu
+, liberation_ttf, curl, util-linux, xdg-utils, wget
+## Arch Linux.
+, flac, harfbuzz, icu, libpng, libopus, snappy, speechd
+## Gentoo
+, bzip2, libcap
```
```suggestion
, liberation_ttf, curl, util-linux, xdg-utils, wget
, flac, harfbuzz, icu, libpng, libopus, snappy, speechd
, bzip2, libcap
```
If we are not going to link the decision or the reasons why the other distros added those, then we might as well omit them.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/wavebox/default.nix:
```diff
@@ -1,94 +1,153 @@
+
+, gsettings-desktop-schemas
+
+# For video acceleration via VA-API (--enable-features=VaapiVideoDecoder)
+, libvaSupport ? true, libva
```
libva is tiny. Why make it an option?

flexiondotorg replied:
This is a hangover of basing the revised Wavebox package on the Google Chrome package in Nixpkgs. The Google Chrome package has this option, but I can drop it.

flexiondotorg replied:
The origin of having the option to disable VA-API is to accommodate systems where VA-API doesn't work reliably.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/wavebox/default.nix:
```diff
@@ -1,94 +1,153 @@
+# For video acceleration via VA-API (--enable-features=VaapiVideoDecoder)
+, libvaSupport ? true, libva
+
+# For Vulkan support (--enable-features=Vulkan)
```
The comment suggests that this enabled vulkan support trough a libary but the hook just adds our driver directory, not any libraries on itself.

flexiondotorg replied:
This is also a hangover of basing the revised Wavebox package on the Google Chrome package in Nixpkgs.

JohnRTitor replied:
Can this be removed?

flexiondotorg replied:
The origins of having the option to disable VA-API is to accommodate systems where VA-API doesn't work reliably.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/wavebox/default.nix:
```diff
@@ -1,94 +1,153 @@
+, alsa-lib, libXdamage, libXtst, libXrandr, libxshmfence, expat, cups
+, dbus, gtk3, gtk4, gdk-pixbuf, gcc-unwrapped, at-spi2-atk, at-spi2-core
+, libkrb5, libdrm, libglvnd, mesa
+, libxkbcommon, pipewire, wayland # ozone/wayland
```
```suggestion
, libxkbcommon, pipewire, wayland
```

SuperSandro2000 reviewed (commented):
Putting every string behind a variable if even if not necessary, just makes the code hard to read and understand.

jnsgruk commented:
Sorry I was late to this, I just saw it come up after SuperSandro's review. Might be worth adding a commit at the end that reformats using `nixfmt-rfc-style` while we're here.

flexiondotorg committed:
wavebox: apply suggestions from code review

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

JohnRTitor commented:
Error when running nixpkgs-review.
```
error:
       … while calling the 'listToAttrs' builtin

         at /nix/store/fg7dh6aandk5iy5hmfr5n6n8nqggsr93-nixpkgs-review-2.10.5/lib/python3.12/site-packages/nixpkgs_review/nix/evalAttrs.nix:46:1:

           45|
           46| listToAttrs (concatMap getProperties attrs)
             | ^
           47|

       … while calling the 'concatMap' builtin

         at /nix/store/fg7dh6aandk5iy5hmfr5n6n8nqggsr93-nixpkgs-review-2.10.5/lib/python3.12/site-packages/nixpkgs_review/nix/evalAttrs.nix:46:14:

           45|
           46| listToAttrs (concatMap getProperties attrs)
             |              ^
           47|

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: evaluation aborted with the following error message: 'lib.customisation.callPackageWith: Function called without required argument "addOpenGLRunpath" at /home/masum/.cache/nixpkgs-review/pr-324822-1/nixpkgs/pkgs/applications/networking/instant-messengers/wavebox/default.nix:34'
```

flexiondotorg committed:
wavebox: bump version to 10.127.10-2

flexiondotorg committed:
wavebox: complete the removal of $appname variable

flexiondotorg committed:
wavebox: use callPackage from all-packages.nix; wavebox is not a Qt5 app

flexiondotorg committed:
wavebox: replace addOpenGLRunpath with addDriverRunpath

flexiondotorg committed:
wavebox: update comment about libvaSupport

flexiondotorg committed:
wavebox: format with nixfmt-rfc-style

flexiondotorg commented:
Rebased against master, updated to Wavebox 10.127.10-2, recent Google Chrome updates in Nixpkgs incorporated and reformatted with `nixfmt` as per @jnsgruk suggestion.

jnsgruk commented:
Result of `nixpkgs-review pr 324822` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>wavebox</li>
  </ul>
</details>

jnsgruk reviewed (approved):
LGTM, assuming @SuperSandro2000 is content with your commentary on the LD_LIBRARY_PATH hack.

I built this locally and ran it, too.

JohnRTitor reviewed (dismissed):
The commits need to be reworded correctly. They should start with `wavebox: ` indicating this change affects the wavebox package.

jnsgruk commented:
Good catch.

JohnRTitor commented:
Result of `nixpkgs-review pr 324822` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>wavebox</li>
  </ul>
</details>
